src/sinks/splunk_hec/common/service.rs

```diff
@@ -38,19 +96,69 @@ impl Service<HecRequest> for HecService {
+                Poll::Ready(Some(permit)) => {
+                    if self.current_ack_slot.replace(permit).is_some() {
+                        Poll::Ready(Err("poll_ready called after a successful call".into()))
+                    } else {
+                        Poll::Ready(Ok(()))
+                    }
+                }
+                Poll::Ready(None) => Poll::Ready(Err(
+                    "Indexer acknowledgements semaphore unexpectedly closed".into(),
+                )),
+                Poll::Pending => Poll::Pending,
+            }
+        }
     }
 
     fn call(&mut self, req: HecRequest) -> Self::Future {
         let mut http_service = self.batch_service.clone();
+        let ack_finalizer_tx = self.ack_finalizer_tx.clone();
+        let ack_slot = self.current_ack_slot.take();
+
         Box::pin(async move {
             http_service.ready().await?;
             let events_count = req.events_count;
             let events_byte_size = req.events_byte_size;
             let response = http_service.call(req).await?;
             let event_status = if response.status().is_success() {
-                EventStatus::Delivered
+                if let Some(ack_finalizer_tx) = ack_finalizer_tx {
+                    let _ack_slot = ack_slot.expect("poll_ready not called before invoking call");
+                    let body = serde_json::from_slice::<HecAckResponseBody>(response.body());
+                    match body {
+                        Ok(body) => {
+                            if let Some(ack_id) = body.ack_id {
+                                let (tx, rx) = oneshot::channel();
+                                match ack_finalizer_tx.send((ack_id, tx)).await {
+                                    Ok(_) => rx.await.unwrap_or(EventStatus::Rejected),
+                                    // If we cannot send ack ids to the ack client, fall back to default behavior
+                                    Err(_) => {
+                                        emit!(&SplunkIndexerAcknowledgementUnavailableError);
+                                        EventStatus::Delivered
+                                    }
+                                }
+                            } else {
+                                // Default behavior if indexer acknowledgements is disabled on the Splunk server
+                                EventStatus::Delivered
+                            }
+                        }
+                        Err(error) => {
+                            // This may occur if Splunk changes the response format in future versions.
+                            emit!(&SplunkResponseParseError { error });
+                            EventStatus::Delivered
+                        }
+                    }
+                } else {
+                    // Default behavior if indexer acknowledgements is disabled by configuration
+                    EventStatus::Delivered
+                }
```

Comment:
Just a thought, non-blocking: I wonder if you could take advantage of the various optional values being used here to actually avoid returning `EventStatus::Delivered` at the end of every branch except for the rejected case, and boil it down to something like `ack_finalizer_tx.map(....).map(...).unwrap_or(EventStatus::Delivered)`.

Reply:
Sure, will toy around with this -- think I ran into some issues last time I'd tried simplifying the many branches but forget exactly what they were.